Load sysctl config file on startup

amclain opened the pull request:
Depends on nerves-project/nerves_system_br#570
Related to elixir-toolshed/toolshed#81

~~Out of the box on a Nerves target, `net.ipv4.ping_group_range` is set to `1 0` which causes an `:eacces` error when trying to open a socket to send an ICMP ping. This PR adjusts this parameter when a Nerves application starts up so that all users can send a ping.~~

This PR loads the `sysctl` config file on startup. The `sysctl.conf` file in nerves-project/nerves_system_br#570 allows all user groups to send an ICMP ping.

fhunleth commented:
Thanks for sending the PR. 

This isn't the first time that setting a sysctl has come up and I'm leaning towards just loading the whole `/etc/sysctl.conf` file in nerves_runtime. The reason is that merging this would set a precedent for setting sysctl's in `nerves_runtime`. We'll probably want to add more sysctl's in the future, and then at some point will want to make them optional. So, how about this:

1. `nerves_runtime` runs `sysctl -p` to load all values from `/etc/sysctl.conf`
2. `nerves_system_br` gets updated to have a default `sysctl.conf` with the `net-ipv4.ping_group_range` setting. The nice part is that it can get a comment that says why this is nice to have for anyone wondering.
3. The next Nerves systems will have the sysctl value set and they can add more to the `sysctl.conf` if they want.
4. Users can add a custom `sysctl.conf` in their rootfs overlay if they don't like the default set.

amclain commented:
👍 I've updated this PR to the `sysctl -p` method, and added the default `sysctl.conf` file to `nerves_system_br` in nerves-project/nerves_system_br#570.

fhunleth commented on lib/nerves_runtime/application.ex:
```diff
@@ -20,6 +20,12 @@ defmodule Nerves.Runtime.Application do
 
     target = Nerves.Runtime.target()
 
+    _ =
+      if target != "host" do
+        # Load sysctl config file (/etc/sysctl.conf).
+        System.cmd("sysctl", ["-p"])
+      end
+
```
If there's a failure, it would be nice to log the error. I also have a small preference to make the paths explicit even though I'm not too worried about this breaking. Perhaps a helper function like this:

```
  # call with load_sysctl_conf("/etc/sysctl.conf")

 defp load_sysctl_conf(path) do
    if File.exists?(path) do
      case System.cmd("/sbin/sysctl", ["-p", path]) do
        {_, 0} ->
          :ok
        {reason, _non_zero_exit} ->
          Logger.warn("Failed to run sysctl on #{path}: #{inspect(reason)}")
          false
      end
    else
      # Don't run sysctl if the conf file doesn't exist.
      :ok
    end

amclain replied:
👍 No problem! I've molded your recommendation into the style of the existing `try_entropy_generator` function so the codebase looks consistent, if that looks good to you.

fhunleth commented:
I rebased and removed the warning if there's no `sysctl.conf`. At the moment, it's only the couple of us who have a `sysctl.conf` file and I don't want anyone who doesn't have one to get confused that they should.
